platform/kube/admit/admit.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admit
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1alpha1"
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+	admissionClient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1alpha1"
+
+	"istio.io/pilot/adapter/config/crd"
+	"istio.io/pilot/model"
+)
+
+// A cluster-unique (i.e. random) suffix should be added to each
+// default below when testing in a shared cluster to avoid collisions.
+
+const (
+	// DefaultAdmissionHookConfigName is the default name for the
+	// ExternalAdmissionHookConfiguration.
+	DefaultAdmissionHookConfigName = "pilot-config"
+
+	// DefaultAdmissionHookName is the default name for the
+	// ExternalAdmissionHooks.
+	DefaultAdmissionHookName = "pilot.config.istio.io"
+
+	// DefaultAdmissionServiceName is the default service of the
+	// validation webhook.
+	DefaultAdmissionServiceName = "istio-pilot-config"
```

Comment:
should this be pilot itself? seems unnecessary to allocate a service just for validation.

Reply:
Yes, though we'll need a new port since this is HTTPS. We'll also need to temporary use a different service name as a workaround for the issue described [here](https://github.com/kubernetes/kubernetes/issues/49987#issuecomment-319739227), but that should eventually go away (k8s 1.8+).

Reply:
I see, then it makes sense but deserves a note in the comment.